File: Credit/credit.py
# get method get_int from module cs50
from cs50 import get_int


def main():
    card_input = get_input()
    card_num = card_input[0]
    card_type = card_input[1]

    if not card_type == "INVALID":
        # put card numbers into int list
        card = [int(c) for c in card_num]
        card_type = check_start(card, card_type)

        if not card_type == "INVALID":
            if not check_sum(card):
                card_type = "INVALID"

    print(card_type)


# apply algorithm on card numbers
def check_sum(card):
    tot_sum = 0

    # get every other digit from last 2nd, each multiply by 2
    for i in range(len(card)-2, -1, -2):
        # print("------")
        #print("i: ", i)
        tot_sum += ((card[i]*2) // 10) + ((card[i]*2) % 10)
        #print("tot_sum: ", tot_sum)

    # get the rest of the digits and add together
    for j in range(len(card)-1, -1, -2):
        # print("------")
        #print("j: ", j)
        #print("card[j]: ", card[j])
        tot_sum += card[j]

    #print("tot_sum: ", tot_sum)
    # check last digit is 0
    while True:
        tot_sum %= 10
        if tot_sum < 10 and tot_sum == 0:
            return True
        else:
            return False


# check the starting numbers of the card
def check_start(card, card_type):
    #print("card[0]: ", card[0])
    #print("card[1]: ", card[1])
    #print("card_type: ", card_type)
    #print("len(card_type): ", len(card_type))

    # check AMEX
    if card[0] == 3:
        if card[1] == 4 or card[1] == 7:
            if card_type == "AMEX":
                return "AMEX"

    # check MASTERCARD
    if card[0] == 5:
        if card[1] >= 1 or card[1] <= 5:
            if card_type == "MASTERCARD" or card_type == "Master/Visa":
                return "MASTERCARD"

    # check VISA
    if card[0] == 4:
        if card_type == "VISA" or card_type == "Master/Visa":
            return "VISA"

    return "INVALID"


# check input and return num
def get_input():
    card_type = ""
    while True:
        card_num_int = get_int("Number: ")
        card_num = str(card_num_int)

        if card_num_int > 0:
            #print("card num > 0")
            if len(card_num) == 15:
                #print("card length == 15")
                card_type = "AMEX"
            elif len(card_num) == 13:
                #print("card length == 13")
                card_type = "VISA"
            elif len(card_num) == 16:
                #print("card length == 16")
                card_type = "Master/Visa"
            else:
                #print("card length == not valid")
                card_type = "INVALID"

        if card_type:
            #print ("cardtype not null")
            break

    #print("card type: ", card_type)
    #print("card num: ", card_num)
    return (card_num, card_type)


if __name__ == "__main__":
    main()

File: credit/credit.c
#include <stdio.h>
#include <cs50.h>

int card_length;
int first_digit;
int first_two_digits;
int digits_from_back[16];
bool checksum;

long get_input(void);
void card_validity(long card_num);
string get_type(long card_num);
void get_details(long card_num);
bool check_sum(void);


int main(void)
{
    long card_num = get_input();

    card_validity(card_num);
}

// get credit card number from user
long get_input(void)
{
    long input;

    do
    {
        input = get_long("Number: ");
    }
    while (input < 0);

    //printf("Input: %ld\n", input);
    return input;
}

// start checks on validity
void card_validity(long card_num)
{
    //printf("in card_validity...\n");

    string card_type = get_type(card_num);

    printf("%s\n", card_type);
}

// get card type
string get_type(long card_num)
{
    //printf("in get_type...\n");

    string card_type;

    get_details(card_num);
    checksum = check_sum();

    if (card_length == 15 && (first_two_digits == 34 || first_two_digits == 37) && checksum)
    {
        card_type = "AMEX";
    }
    else if (card_length == 16 && (first_two_digits >= 51 && first_two_digits <= 55) && checksum)
    {
        card_type = "MASTERCARD";
    }
    else if ((card_length == 13 || card_length == 16) && first_digit == 4 && checksum)
    {
        card_type = "VISA";
    }
    else
    {
        card_type = "INVALID";
    }

    //printf("card_type: %s\n", card_type);
    return card_type;
}

bool check_sum(void)
{
    //printf("in check_sum...\n");

    int odd = 0;
    int even = 0;
    int total = 0;

    for (int i = 0; i < card_length; i += 2)
    {
        odd += digits_from_back[i];
    }

    for (int i = 1; i < card_length; i += 2)
    {
        int temp = digits_from_back[i];
        temp *= 2;
        even += temp / 10 + temp % 10;
    }

    total = odd + even;

    if (total % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

// get length and digit details
void get_details(long card_num)
{
    //printf("in get_details...\n");

    long remain_length = card_num;

    card_length = 0;

    while (remain_length > 0)
    {
        digits_from_back[card_length] = remain_length % 10;

        if (remain_length < 10)
        {
            first_digit = remain_length;
        }
        else if (remain_length < 100)
        {
            first_two_digits = remain_length;
        }

        remain_length /= 10;

        card_length++;
        //printf("card_length: %i\n", card_length);
    }
}